text/0567-use-and-resources.md

```diff
@@ -0,0 +1,953 @@
+{{else}}
+  {{#each this.products.data as |product|}}
+    <ProductCard @product={{product}} />
+  {{/each}}
+{{/if}}
+```
+
+## Motivation
+
+Ember Octane introduced autotracking, which is a general-purpose reactivity
+system that redefines the way that state flows throughout Ember applications.
+Native getters, helpers, and modifiers have all been integrated with
+autotracking, and collectively these changes have made state flow more
+declarative and targeted, and generally have been well received by early
+adopters.
+
+One of the pieces of feedback we've received, however, is that there are some
+gaps in the this reactivity sytem. Some example use cases include:
+
+- Triggering asynchronous requests based on changing arguments, and consuming
+  the results in templates and getters in a natural way.
+- Controlling external libraries such as Leaflet and Three.js using components
+  and the rendering tree, rather than manual library code.
+- More generally, controlling state outside of Ember itself. A great example is
+  [TurboPatent](https://turbopatent.com/), an Electron app which needs to sync
+  the state of the app with the state of the _native_ menu bar, outside of the
+  document and Ember's rendering layer entirely.
+
+These use cases were handled sporadically in a number of different ways in
+Ember Classic. Users could use component lifecycle hooks, observers, and
+external libraries like `ember-concurrency` and `ember-async-await-helper` to
+solve these problems, but there was not a good general solution that fit all of
+them.
+
+Fundamentally, the issues come down to mixing _declarative_ code (like getters
+and helpers, where tracked properties are used to produce a new result based on
+their values) and _imperative_ code (like component lifecycle hooks Ember
+Classic, where a the user wants to run some code that does not return a value
+whenever something else changes).
+
+This RFC proposes a new primitive, _Resources_, built specifically for this
+general problem of mixing imperative and declarative code; one that will
+complement the existing Octane programming model, and improve its ergonomics
+overall.
+
+### Case Study: Triggering Async
+
+To demonstrate the gap that Resources fill, we can take a look at a fairly
+common use case that users of `ember-concurrency` tasks have had: Rerunnning a
+task whenever a component argument changes.
+
+In Ember Classic, it would be straightforward to set this up:
+
+```js
+export default Component.extend({
+  products: task(function*(url) {
+    let response = yield fetch(url);
+    let data = yield response.json();
+    return data;
+  }),
+
+  didReceiveAttrs() {
+    this.products.perform(this.searchUrl;)
+  },
+});
+```
+
+With Ember Octane, it is possible to get the same behavior using
+`ember-concurrency-decorators` and `@ember/render-modifiers`, but it is a bit
+more convoluted:
+
+```js
+export default class SearchResults extends Component {
+  @task
+  products*(url) {
```

Comment:
To match the usage in the template below, I think this should be changed

```suggestion
  loadProducts*(url) {
```

Reply:
The asterisk should come before the function name: `*loadProducts(url) {`